### ISSUE TITLE
Synchronous vs Asynchronous services, and difference between ok and completed

### ISSUE THREAD

**Issue by erikbosch** [COVESA/vehicle_service_catalog#28]
In the current [error file](https://github.com/COVESA/vehicle_service_catalog/blob/master/vsc-error.yml) we have a few positive error codes, like `ok`, `in_progress` and `completed`. In the existing [seat service](https://github.com/COVESA/vehicle_service_catalog/blob/master/comfort-service.yml#L186) we sometimes use all three 3 of them:

`range: $ in_set("ok", "completed", "in_progress", "interrupted")`

In the current error list the difference between e.g.  `ok` and `completed` is not clearly stated, so it is unclear when they shall be used and what they mean. This brings up some discussion topics:

- Do we need a mechanism in VSC (formal or informal) to specify if a service (implementation) shall be asynchronous or synchronous, and/or if a service shall return as soon as command has been accepted, or wait until operation is fully completed? Or is that up to the implementation framework? E.g. something similar to a layer-concept that specifies if the specific implementation of a service is sync or async?
- Do we expect an implementation (on server side) to be able to return multiple error codes. Is it e.g. intended that server side service implementation actually shall report `in_progress`, or is that just something that the framework might use as an intermediate value?
- What is the semantic difference between `ok` and `completed`? Is it relevant that a method use both?

Would it possibly be useful to have more details on when error codes shall be used, maybe something similar like: https://grpc.github.io/grpc/cpp/md_doc_statuscodes.html

**Comment by gunnarx** [comment]
> Do we need a mechanism in VSC (formal or informal) to specify if a service (implementation) shall be asynchronous or synchronous

Yes I think we need to agree and specify it somewhere, but it might still be part of "deployment model".
  In either case I'd like to recommend that this definition is deferred to as late as possible.  In other words it should be possible to define an abstract "API" (the method signatures) without being forced to specify this, as I think it is a detail of deployment whether an operation is sync or async.

**Comment by gunnarx** [comment]
> Do we expect an implementation (on server side) to be able to return multiple error codes. Is it e.g. intended that server side service implementation actually shall report in_progress, or is that just something that the framework might use as an intermediate value?

Update: The **Errors** concept should now include this, at least in #41 that is being discussed.

... but if you are talking about the **Return Value** concept, the TERMS.md proposes that it may only have one type. 
But of course since the type is arbitrary it's theoretically possible to encode multiple values in it.

**Comment by gunnarx** [comment]
> What is the semantic difference between ok and completed? Is it relevant that a method use both?

Relevant question, about the [example error file](https://github.com/COVESA/vehicle_service_catalog/blob/master/vsc-error.yml).  It's not a language definition question, but it is something that ought to be fixed(?) in the error example I guess.

**Comment by gunnarx** [comment]
> Would it possibly be useful to have more details on when error codes shall be used, maybe something similar like: https://grpc.github.io/grpc/cpp/md_doc_statuscodes.html

Well I agree, as part of the example error file.  I'd say predefined errors is not part of the VSC core specification, but it is of course possible to propose something as a complementary specification.

_Sidebar:  It's really a tricky thing to communicate to casual readers what is part of a firm specification and what is just part of examples.  There is plenty of opportunity for misunderstanding (also within the core team I guess)._

**Comment by gunnarx** [comment]
Since this is a question for the definition of the language/interface-model that is now [IFEX](https://github.com/COVESA/ifex) project, the issue will be transferred from Vehicle Service Catalog project to IFEX project.
